layer_factory/assistant_layer/mem_bound_warning.cpp

```diff
@@ -0,0 +1,210 @@
+                SetBindMemoryState<VkBuffer>(bindInfo.pBufferBinds[j].buffer, sparse_binding.memory);
+            }
+        }
+        for (uint32_t j = 0; j < bindInfo.imageOpaqueBindCount; j++) {
+            for (uint32_t k = 0; k < bindInfo.pImageOpaqueBinds[j].bindCount; k++) {
+                auto sparse_binding = bindInfo.pImageOpaqueBinds[j].pBinds[k];
+                SetBindMemoryState<VkImage>(bindInfo.pImageOpaqueBinds[j].image, sparse_binding.memory);
+            }
+        }
+        for (uint32_t j = 0; j < bindInfo.imageBindCount; j++) {
+            for (uint32_t k = 0; k < bindInfo.pImageBinds[j].bindCount; k++) {
+                auto sparse_binding = bindInfo.pImageBinds[j].pBinds[k];
+                SetBindMemoryState<VkImage>(bindInfo.pImageBinds[j].image, sparse_binding.memory);
+            }
+        }
+    }
+
+    return VK_SUCCESS;
+}
+
+VkResult MemBoundWarning::PostCallGetSwapchainImagesKHR(VkDevice device, VkSwapchainKHR swapchain, uint32_t *pSwapchainImageCount,
+                                                        VkImage *pSwapchainImages, VkResult result) {
+    assert(swapchain);
+    auto images = swapchain_image_map[swapchain];
+
+    if (*pSwapchainImageCount > images.size()) images.resize(*pSwapchainImageCount);
+
+    if (pSwapchainImages) {
+        for (uint32_t i = 0; i < *pSwapchainImageCount; i++) {
+            if (images[i] != VK_NULL_HANDLE) continue;  // Already retrieved image
+            images[i] = pSwapchainImages[i];
+        }
+    }
+
+    return VK_SUCCESS;
+}
+
+// Get memory associated with the handle being destroyed then find handle set using this memory and erase it
+template <typename T>
+void MemBoundWarning::RemoveBindedMemory(T handle) {
+    assert(handle);
+    uint64_t handle_64 = HandleToUint64(handle);
+    auto map_pair = handle_memory_map.find(handle_64);
+    if (map_pair == handle_memory_map.end()) return;  // this will happend if handle was never Binded to memory...prevents segfault
+    auto memory_set = map_pair->second;
+
+    // remove all memory in the memory set. If it is not sparse memory, there will only be one item in set
+    for (VkDeviceMemory memory : memory_set) {
+        auto map_pair = memory_handle_map.find(memory);
+        if (map_pair == memory_handle_map.end())
+            continue;  // this will happend if handle was never Binded to memory...prevents segfault
+        auto &handle_set = map_pair->second;
+        handle_set.erase(handle_64);
+    }
+}
+
+void MemBoundWarning::PostCallDestroySwapchainKHR(VkDevice device, VkSwapchainKHR swapchain,
+                                                  const VkAllocationCallbacks *pAllocator) {
+    assert(swapchain);
+
+    auto map_pair = swapchain_image_map.find(swapchain);
+    if (map_pair == swapchain_image_map.end())
+        return;  // this will happend if handle was never Binded to memory...prevents segfault
+    auto images = map_pair->second;
```

Comment:
It looks like this is making a copy. Did you intend to use a reference instead?

Reply:
In this case, since the images are only read, and not written to. A reference may not be necessary. However, making it a reference improves memory efficiency right? I will also make it a constant since it is read only. 
Thoughts?

Reply:
Yes, I would make this a const reference to avoid the cost of copying a std::vector.